subsys/net/lib/nrf_cloud/src/nrf_cloud_jwt.c

```diff
@@ -8,12 +8,161 @@
 #include <zephyr/kernel.h>
 #include <zephyr/logging/log.h>
 #include <modem/modem_jwt.h>
+#if defined(CONFIG_NRF_MODEM_LIB)
 #include <nrf_modem_at.h>
+#endif
+#include <string.h>
+#include <net/nrf_jwt.h>
+#include <zephyr/sys/base64.h>
+#include <psa/crypto.h>
+
+/* For testing */
+#define READ_FROM_INC_FILE 0
+#if READ_FROM_INC_FILE
+#include CONFIG_NRF_CLOUD_CERTIFICATES_FILE
+#endif
 
 #define GET_TIME_CMD "AT%%CCLK?"
 
 LOG_MODULE_REGISTER(nrf_cloud_jwt, CONFIG_NRF_CLOUD_LOG_LEVEL);
 
+#define PRV_KEY_SZ (NRF_JWT_KEY_DER_SZ)
+#define PRV_KEY_DER_SZ (138)
+#define PRV_KEY_PEM_SZ (256)
+#define PRV_KEY_DER_START_IDX (36)
+
+#if defined(CONFIG_NRF_CLOUD_JWT_GEN_METHOD_PSA)
+#define JWT_KEY_ID CONFIG_NRF_CLOUD_JWT_PSA_KEY_ID
+#else
+#define JWT_KEY_ID PSA_KEY_ID_NULL
+#endif
+
+static void remove_newlines(char *const str)
+{
+	size_t new;
+	size_t len = strlen(str);
+
+	for (size_t old = new = 0; old < len; ++old) {
+		if (str[old] != '\n') {
+			str[new++] = str[old];
+		}
+	}
+
+	str[new] = '\0';
+}
+
+#define BEGIN_PRV_KEY "-----BEGIN PRIVATE KEY-----"
+#define END_PRV_KEY "-----END PRIVATE KEY-----"
+static int strip_non_key_data(char *const str)
+{
+	char *start;
+	char *end;
+	size_t new_len;
+
+	/* Find and remove end string */
+	end = strstr(str, END_PRV_KEY);
+	if (end == NULL) {
+		return -EINVAL;
+	}
+	*end = '\0';
+
+	/* Find and remove begin string */
+	start = strstr(str, BEGIN_PRV_KEY);
+	if (start == NULL) {
+		return -EINVAL;
+	}
+
+	start += strlen(BEGIN_PRV_KEY);
+	new_len = strlen(start);
+
+	/* Move key data to the front */
+	memmove(str, start, new_len);
+	str[new_len] = '\0';
+
+	remove_newlines(str);
+
+	return 0;
+}
+
+static int open_key(psa_key_id_t key_id, psa_key_handle_t *const key_h_out)
+{
+#if defined(CONFIG_NRF_CLOUD_JWT_GEN_METHOD_PSA)
+	int ret = psa_crypto_init();
+
+	if (ret != PSA_SUCCESS) {
+		LOG_ERR("psa_crypto_init() failed, error: %d", ret);
+		return -EIO;
+	}
+
+	ret = psa_open_key(key_id, key_h_out);
+
+	if (ret != PSA_SUCCESS) {
+		LOG_ERR("psa_open_key() failed, error: %d", ret);
+		return -EACCES;
+	}
+
+	return 0;
+#else
+	return -ENOTSUP;
+#endif
+}
+
+static int get_key_from_cred(struct jwt_data *const jwt)
+{
+	static char pem[PRV_KEY_PEM_SZ];
+	static char der[PRV_KEY_DER_SZ];
+	size_t pem_sz = sizeof(pem);
+	size_t der_sz;
+	int err;
+
+#if READ_FROM_INC_FILE
+	/* If the file is included here, just read it */
+	pem_sz = strlen(private_key) + 1;
+
+	if (pem_sz > PRV_KEY_PEM_SZ) {
+		return -EMSGSIZE;
+	}
+	memcpy(pem, private_key, pem_sz);
+#else
+	/* Get the private key from protected storage */
```

Comment:
Worth noting: Whether it comes from protected storage depends on the backend being used with the TLS credentials module